src/v/pandaproxy/schema_registry/util.h

```diff
@@ -47,4 +55,39 @@ result<schema_definition> make_schema_definition(std::string_view sv) {
       ss::sstring{str_buf.GetString(), str_buf.GetSize()}};
 }
 
+///\brief The first invocation of one_shot::operator()() will invoke func and
+/// wait for it to finish. Concurrent invocatons will also wait.
+///
+/// On success, all waiters will be allowed to continue. Successive invocations
+/// of one_shot::operator()() will return ss::now().
+///
+/// If func fails, waiters will receive the error, and one_shot will be reset.
+/// Successive calls to operator()() will restart the process.
+class one_shot {
+    enum class state { empty, started, available };
+
+public:
+    explicit one_shot(ss::noncopyable_function<ss::future<>()> func)
+      : _func{std::move(func)} {}
+    ss::future<> operator()() {
+        if (likely(_state == state::available)) {
+            return ss::now();
+        }
+        if (std::exchange(_state, state::started) == state::empty) {
+            _prom_started = {};
+            (void)_func().then_wrapped([this](ss::future<> f) {
+                _state = f.failed() ? state::empty : state::available;
+                f.failed() ? _prom_started.set_exception(f.get_exception())
+                           : _prom_started.set_value();
+            });
+        }
+        return _prom_started.get_shared_future();
+    }
+
+private:
+    ss::noncopyable_function<ss::future<>()> _func;
+    ss::shared_promise<> _prom_started{};
```

Comment:
Yeah, I struggled to avoid a shared_promise here. This suggestion could also work to know when one_shot can be destroyed.

Reply:
I think you could also get rid of the one_shot utility completely and use a custom built solution for this case. the following illustrates the idea, but you could probably wrap it up in some sugar:

```
ss::semaphore topic_create_sem{1};
request_handler(...) {
return ss::with_semaphore(topic_create_sem, 1, [this] {
   if (topic_create_sem.count() == 0) {
      // do create topic
      topic_create_sem.signal(semaphore::max_counter());
   }
   // handle request
});
}
```

the basic idea is: initially the semaphore is a mutex so exactly one fiber will be allowed to acquire the semaphore. that fiber can discover that it is the chosen one and create the topic. before returning, it sets the semaphore to be max counter which ensures that no future requester ever blocks on the semaphore.

this isn't completely background/async like one_shot, but it would only effect the the very first requests

Reply:
Everybody waits for it anyway. I like this approach. I hadn't thought to just set the semaphore to max, I think I dismissed the idea trying to choose an arbitrary number of units. I'll change it, it's a simpler solution, and solves the lifetime issue.

Reply:
cool. I think its a little more complicated than the example I gave. the main complication is the first fiber detecting its the first. the above illustrates the idea, but it assumes that the continuation runs inline with semaphore acquisition and thus checking for zero is guaranteed to happen without interrupt wrt decrementing the semaphore. so we might need to use try_lock or something slightly more low-level to guarantee correctness under all scheduling scenarios.

Reply:
Reworked `one_shot` around a `semaphore`.